### PR TITLE
Add MediaResponse object to support pagination headers

### PR DESCRIPTION
**Pull request by roborourke**
This is a first pass at adding support for WordPress 5.8 requiring pagination headers for the media library.

Related to #38

### CODE REVIEW

File: inc/Provider.php [hunk 1]
```diff
@@ -220,7 +222,7 @@ final public function remote_request( string $url, array $args ) : string {
 			);
 		}
 
-		return $response_body;
+		return $response['http_response'];
```
Comment (tfrommen) [inline inc/Provider.php:225]: Is this always set?

File: inc/LocalProvider.php [hunk 2]
```diff
@@ -18,11 +18,11 @@ public function get_name() : string {
 		return apply_filters( 'amf/local_provider/name', __( 'Local Media', 'asset-manager-framework' ) );
 	}
 
-	protected function request( array $args ) : MediaList {
+	protected function request( array $args ) : MediaResponse {
 		// Call the default core attachment query AJAX handler.
 		// This will return JSON and exit before the return statement below.
 		wp_ajax_query_attachments();
-		return new MediaList();
+		return new MediaResponse( new MediaList(), 1, 1 );
```
Comment (tfrommen) [inline inc/LocalProvider.php:25]: I know this line is not actually executed, but I'm wondering if we want to provide some defaults to the `MediaResponse` constructor? Assuming an empty list, which means a total of `0` items, and the WordPress default of `40` items per page, we can just do this:

```suggestion
		return new MediaResponse();
```

File: inc/MediaResponse.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+	/**
+	 * The reponse media list.
+	 *
+	 * @var MediaList
+	 */
+	private $media_list;
+
+	public function __construct( MediaList $media_list, int $total, int $per_page ) {
+		$this->media_list = $media_list;
```
Comment (tfrommen) [inline inc/MediaResponse.php:22]: I suggest providing defaults for this to allow for easier usage in failure/empty case:

```suggestion
	public function __construct( MediaList $media_list = null, int $total = 0, int $per_page = 40 ) {
		$this->media_list = $media_list ?? new MediaList();
```

File: inc/MediaResponse.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+		// Set pagination headers.
+		if ( ! headers_sent() ) {
+			header( sprintf( 'X-WP-Total: %d', $total ) );
+			header( sprintf( 'X-WP-TotalPages: %d', ceil( $total / $per_page ) ) );
+		}
```
Comment (tfrommen) [inline inc/MediaResponse.php:28]: Can we please move this outside the constructor? This is a side-effect, and even more one that doesn't have anything to do with the internal data of this class, really.

I suggest to store both `$total` and `$per_page` as fields on this class, and introduce a new method `send_pagination_headers`, which should get called in the `request_items` method of the `Provider` class, right afterthe `$response` has been received (from `$this->request()`).

**Review by tfrommen — CHANGES_REQUESTED** [review]

File: inc/Provider.php (hunk 1 above)
Comment (roborourke) [reply inc/Provider.php:225]: Yep, at least at this stage - the method throws earlier if it's a `WP_Error`.

File: inc/MediaResponse.php (hunk 4 above)
Comment (roborourke) [reply inc/MediaResponse.php:28]: Sure, I did start that way but thought they're only needed here so tried this to be as terse as possible but you're right.

File: readme.md [hunk 5]
```diff
@@ -119,7 +120,11 @@ class KittenProvider extends Provider {
 			$items[] = $item;
 		}
 
-		return new MediaList( ...$items );
+		return new MediaResponse(
+			new MediaList( ...$items ),
+			count( $kittens ), // Total number of available results.
+			count( $Kittens )  // Number of items requested per page.
```
[the remark this answers is not in the document]
Comment (roborourke) [reply readme.md:126]: lol 🤦‍♂️

File: inc/MediaResponse.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	public function get_items() : MediaList {
+		return $this->media_list;
+	}
+
+	public function get_total() : int {
+		return $this->total;
+	}
+
+	public function get_total_pages() : int {
+		return (int) ceil( $this->total / $this->per_page );
```
[the remark this answers is not in the document]
Comment (roborourke) [reply inc/MediaResponse.php:58]: I don't really agree - per page is a value that should already be readily available, so why force implementors to do this calculation too? Unsplash for instance returns `X-Total` and `X-Per-Page` headers in responses. This feels like a more straightforward way to implement it to me.

We could throw an exception if it's 0, given it'll fatal anyway when trying to divide by 0.

For AMF WordPress -1 is a value specific to WP and it should be handled there IMO. so we could throw an exception for negative values too.

Was there anything else that doesn't make sense to you?

[the remark this answers is not in the document]
Comment (tfrommen) [reply inc/MediaResponse.php:58]: Just to clarify, when I said that not all of this makes sense, I meant my own examples of `0` or negative numbers. 🙂

[the remark this answers is not in the document]
Comment (roborourke) [reply inc/MediaResponse.php:58]: Gotcha! No worries. I'll add in the exceptions and checks.

**Review by tfrommen — APPROVED** [review]
Awesome.

**Pull request merged**